src/t8_schemes/t8_scheme_common_cxx.hxx

```diff
@@ -0,0 +1,631 @@
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with t8code; if not, write to the Free Software Foundation, Inc.,
+  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+*/
+
+/** \file t8_scheme_common_cxx.hxx
+ * We provide some functions that are useful across schemes.
+ */
+
+#ifndef T8_SCHEME_COMMON_CXX
+#define T8_SCHEME_COMMON_CXX
+
+#include <t8_element_cxx.hxx>
+#include <t8_element.h>
+#include <sc_functions.h>
+
+class t8_scheme_common_c: public t8_eclass_scheme_c {
+ private:
+  sc_mempool_t *mempool;
+
+ public:
+  /** Destructor for all default schemes */
+  virtual ~t8_scheme_common_c ()
+  {
+    T8_ASSERT (mempool != NULL);
+    SC_ASSERT (mempool->elem_count == 0);
+    sc_mempool_destroy (mempool);
+  }
+  t8_scheme_common_c (t8_eclass_t eclass_in, int elem_size)
+  {
+    element_size = elem_size;
+    mempool = sc_mempool_new (element_size);
+    eclass = eclass_in;
+  }
+
+  /** Use a mempool to get allocate */
+  virtual void
+  t8_element_new (int length, t8_element_t **elem) const
+  {
+    T8_ASSERT (mempool != NULL);
+    T8_ASSERT (0 <= length);
+    T8_ASSERT (elem != NULL);
+    for (int i = 0; i < length; ++i) {
+      elem[i] = (t8_element_t *) sc_mempool_alloc (mempool);
+    }
+  }
+
+  virtual void
+  t8_element_destroy (int length, t8_element_t **elem) const
+  {
+    T8_ASSERT (mempool != NULL);
+    T8_ASSERT (0 <= length);
+    T8_ASSERT (elem != NULL);
+    for (int i = 0; i < length; ++i) {
+      sc_mempool_free (mempool, elem[i]);
+    }
+  }
+#if T8_ENABLE_DEBUG
+  virtual void
+  t8_element_debug_print (const t8_element_t *elem) const
+  {
+    char debug_string[BUFSIZ];
+    t8_element_to_string (elem, debug_string, BUFSIZ);
+    t8_debugf ("%s\n", debug_string);
+  }
+#endif
+
+  virtual void
+  t8_element_general_function (const t8_element_t *elem, const void *indata, void *outdata) const
+  {
+  }
+
+  virtual t8_eclass_t
+  t8_element_child_eclass (int childid) const
+  {
+    return eclass;
+  }
+
+  virtual int
+  t8_element_is_family (t8_element_t **fam) const override
+  {
+    if (t8_element_level (fam[0]) == 0)
+      return 0;
+    t8_element_t *parent, *parent_compare;
+    t8_element_new (1, &parent);
+    t8_element_new (1, &parent_compare);
+    t8_element_parent (fam[0], parent);
+    int num_children = t8_element_num_children (parent);
+    bool is_family = true;
+    for (int ichild = 1; ichild < num_children; ichild++) {
+      t8_element_parent (fam[ichild], parent_compare);
+      if (!t8_element_equal (parent, parent_compare)) {
+        is_family = false;
+        break;
+      }
+    }
+    t8_element_destroy (1, &parent);
+    t8_element_destroy (1, &parent_compare);
+    return is_family;
+  }
+
+  t8_linearidx_t
+  t8_element_linear_id_recursive (t8_element_t *elem, const t8_linearidx_t id, const int level) const
+  {
+    if (t8_element_level (elem) == 0)
+      return id;
```

Comment:
```suggestion
      return 0;
```
otherwise we could call it with ```t8_element_linear_id_recursive(root, 2, 0)```
And the linear id on root level would be 2.

Reply:
I think the original is true, but neither of our suggestions is true.
Only the following is true:
```suggestion
      T8_ASSERT (0<=id && id<t8_element_count_leaves(elem, level);
      return id;
```
The recursive blackbox takes as input an ancestor of the original element, that was at level `level`, and the id of the original element in the refinement of the current ancestor element. As soon, as the ancestor element becomes the root element, the id is the linear id that we are searching for.